bin/propolis-server/src/lib/initializer.rs

```diff
@@ -994,6 +1001,121 @@ impl<'a> MachineInitializer<'a> {
         smb_tables.commit()
     }
 
+    fn generate_bootorder(&self) -> Result<Option<Entry>, Error> {
+        eprintln!(
+            "generating bootorder with order: {:?}",
+            self.boot_order.as_ref()
+        );
+        let Some(boot_names) = self.boot_order.as_ref() else {
+            return Ok(None);
+        };
+
+        let mut order = fwcfg::formats::BootOrder::new();
+
+        let parse_bdf =
+            |pci_path: &propolis_api_types::instance_spec::PciPath| {
+                let bdf: Result<pci::Bdf, Error> =
+                    pci_path.to_owned().try_into().map_err(|e| {
+                        Error::new(
+                            ErrorKind::InvalidInput,
+                            format!(
+                                "Couldn't get PCI BDF for {}: {}",
+                                pci_path, e
+                            ),
+                        )
+                    });
+
+                bdf
+            };
+
+        for boot_entry in boot_names.iter() {
+            if boot_entry.first_boot_only {
+                continue;
+            }
```

Comment:
The machine initializer code only gets called once per instance ensure request--reboot requests don't re-invoke this path. I think we would want to handle the first-boot-only discipline in the fwcfg module instead. 

I haven't looked at the fwcfg code too much, so am spitballing a bit, `struct Fwcfg` impls the `Lifecycle` trait, particularly the `reset` function, so it already has a way to learn about reboots; we could, I think, have a special `Entry::BootOrder` directory entry type that remembers whether it's a one-shot, and on reset, we could go have `Fwcfg` rummage for it and remove the entry if it's marked as one-shot. This would also take care of things over a migration; once the initially-populated entry is removed, it won't be exported anymore, so it won't be present when `FwCfg::import` restores entries.

Reply:
(i've walked back `first_boot_only` for this first PR. i _did_ look at wiring this up through `Lifecycle::reset` and that seems pretty workable, just not necessary right now)